src/language-html/print/children.js

```diff
@@ -0,0 +1,212 @@
+  needsToBorrowPrevClosingTagEndMarker,
+  printClosingTagEndMarker,
+  printClosingTagSuffix,
+  needsToBorrowParentClosingTagStartMarker,
+} = require("./tag.js");
+
+function printChildren(path, options, print) {
+  const node = path.getValue();
+
+  if (forceBreakChildren(node)) {
+    return [
+      breakParent,
+
+      ...path.map((childPath) => {
+        const childNode = childPath.getValue();
+        const prevBetweenLine = !childNode.prev
+          ? ""
+          : printBetweenLine(childNode.prev, childNode);
+        return [
+          !prevBetweenLine
+            ? ""
+            : [
+                prevBetweenLine,
+                forceNextEmptyLine(childNode.prev) ? hardline : "",
+              ],
+          printChild(childPath),
+        ];
+      }, "children"),
+    ];
+  }
+
+  const groupIds = node.children.map(() => Symbol(""));
+  return path.map((childPath, childIndex) => {
+    const childNode = childPath.getValue();
+
+    if (isTextLikeNode(childNode)) {
+      if (childNode.prev && isTextLikeNode(childNode.prev)) {
+        const prevBetweenLine = printBetweenLine(childNode.prev, childNode);
+        if (prevBetweenLine) {
+          if (forceNextEmptyLine(childNode.prev)) {
+            return [hardline, hardline, printChild(childPath)];
+          }
+          return [prevBetweenLine, printChild(childPath)];
+        }
+      }
+      return printChild(childPath);
+    }
+
+    const prevParts = [];
+    const leadingParts = [];
+    const trailingParts = [];
+    const nextParts = [];
+
+    const prevBetweenLine = childNode.prev
+      ? printBetweenLine(childNode.prev, childNode)
+      : "";
+
+    const nextBetweenLine = childNode.next
+      ? printBetweenLine(childNode, childNode.next)
+      : "";
+
+    if (prevBetweenLine) {
+      if (forceNextEmptyLine(childNode.prev)) {
+        prevParts.push(hardline, hardline);
+      } else if (prevBetweenLine === hardline) {
+        prevParts.push(hardline);
+      } else {
+        if (isTextLikeNode(childNode.prev)) {
+          leadingParts.push(prevBetweenLine);
+        } else {
+          leadingParts.push(
+            ifBreak("", softline, {
+              groupId: groupIds[childIndex - 1],
+            })
+          );
+        }
+      }
+    }
+
+    if (nextBetweenLine) {
+      if (forceNextEmptyLine(childNode)) {
+        if (isTextLikeNode(childNode.next)) {
+          nextParts.push(hardline, hardline);
+        }
+      } else if (nextBetweenLine === hardline) {
+        if (isTextLikeNode(childNode.next)) {
+          nextParts.push(hardline);
+        }
+      } else {
+        trailingParts.push(nextBetweenLine);
+      }
+    }
+
+    return [
+      ...prevParts,
+      group([
+        ...leadingParts,
+        group([printChild(childPath), ...trailingParts], {
+          id: groupIds[childIndex],
+        }),
+      ]),
+      ...nextParts,
+    ];
+  }, "children");
+
+  function printChild(childPath) {
```

Comment:
Is it possible to move these functions out?

Reply:
fixed 34c70a9